### PR TITLE
Add composite file uploader

### PR DESCRIPTION
**Pull request by guerler**
This PR adds composite file support to the upload dialog. It includes a modularization of the regular uploader such that common features i.e. ftp popup, settings popup, upload doms, request generation etc. are shared between regular and composite uploads. It also extends the qunit test suite by introducing some initial upload dialog feature tests.

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/datatypes.py [hunk 1]
```diff
@@ -38,6 +38,13 @@ def index( self, trans, **kwd ):
                     dictionary = {}
                     for key in keys:
                         dictionary[key] = elem.get(key)
+                    extension = elem.get('extension')
+                    if extension in datatypes_registry.datatypes_by_extension:
+                        composite_files = datatypes_registry.datatypes_by_extension[ extension ].composite_files
+                        if composite_files:
+                            dictionary['composite_files'] = []
+                            for file in composite_files:
+                                dictionary['composite_files'].append( composite_files[ file ].dict() )
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/datatypes.py]: I think the last 3 lines can be simplified to:

``` python
dictionary['composite_files'] = [_.dict() for _ in composite_files.itervalues()]
```

**Comment by nsoranzo** [comment]
xref: https://trello.com/c/Fyd7pvMY/2460-add-composite-uploads-to-the-new-upload-dialog

**Comment by martenson** [comment]
Manual testing of composite uploads works fine for me.

@guerler Are you considering some kind of hints or introduction to explain what some parts of the dialogue do? I think it is becoming pretty complex modal and it lacks guidance. Maybe in the beginning after opening the modal when the box is empty you could have some basic info in it that disappears when you add first file and appears when you remove it?

**Comment by guerler** [comment]
@martenson I agree, more guidance is always a good idea. Do you mean the regular or composite file uploader part? Maybe we should highlight the info text: 'You can drag&drop files...' better or move it to the main box.

**Comment by martenson** [comment]
I was thinking to use the canvas of the 'regular' upload box as a help space. Possibly the composite too, but there are always files present so it might be confusing.

screenshot 2015-08-18 12 38 01

Also good idea about moving that drag&drop to the box too.

**Comment by guerler** [comment]
@martenson Thx. I thought so, and I think its a good idea. We should definitely do this but I would rather issue a separate PR for this enhancement since it is not directly related to composite uploads and easily separable.

**Comment by martenson** [comment]
:+1: I especially like the modularization, JS seems very clean

This would also benefit from reviews of people more experienced with composite files.

**Comment by dannon** [comment]
+1 Yep, looks like it works to me too, nice job Sam.

**Comment by martenson** [comment]
@guerler I created an issue for the upload help ref'd above.

**Pull request merged**